### ISSUE TITLE
Future roadmap for Jekyll?

### ISSUE THREAD

**Issue by jaredcwhite** [jekyll/jekyll#8085]
To preface my question, I've professionally built and deployed dozens of Jekyll-based site over the last several years for both myself and a number of clients, and generally it's been an awesome experience. I have immense appreciate for the Jekyll project and all its contributors.

That being said…

As I continue to see massive mindshare gains for SSGs on the Javascript side of the fence (Gatsby, Next.js, etc.) as well as SSGs in other languages (Hugo, etc.), I'm finding it increasingly hard to base my future business plans around Jekyll solely on the basis that I love Ruby and Jekyll gets the job done at the present moment.

So my question is: is there an ambitious, overarching roadmap for Jekyll to help guide its development goals and philosophy well into the future? And if not, why not? While it's great that various performance improvements, bug fixes, etc. keep making their way to the project, I've gotten the sense for a while now that community momentum to meaningfully evolve things forward has died down, and that seems like a missed opportunity.

I've been thinking about some of the ways I've "added-on" to Jekyll on my own projects…for example:

* **Components**. I think a big reason why React-based SSGs are catching on is because web development is moving away from "templates" and "partials" and moving towards components. A page is a component, a section is a component, a block of text is a component, a button is a component. Something I've done with Jekyll is create a Liquid tag (called, unsurprisingly, "component") which adds a thin layer of additional functionality on top of the include tag to make it easier to render Markdown content via components within posts and layouts and such, and I've even experimented with shared component libraries that can be used by multiple sites within a monorepo. What's missing in all of this still is a good testing solution—for example, I think a variant of Storybook.js that could render Liquid components and have them look the same in Strorybook as they would rendered via Jekyll would be way cool.

* **Webpack**. I don't use Jekyll's assets tooling at all, I use Webpack to handle all my JS/CSS assets. Webpack standardization means I can switch between Jekyll sites, Rails apps, and even all-JS projects and never change how I add, import, and use frontend packages. While I'm not suggesting Jekyll officially adopt Webpack, I do think an officially-sanctioned `jekyll-webpack` plugin that could be added to a Gemfile with one line and set Webpack up easily just like `webpacker` does for Rails would be extremely beneficial for the community.

* **Third-party APIs**. While it's absolutely true that Jekyll plugins can pull in data from third-party APIs and add content and pages (and I've done so), I feel like Jekyll doesn't really do anything to facilitate that other than provide the basic plugin infrastructure. Projects like Gatsby include support for common APIs like GraphQL as a first-party, officially-supported way of pulling data into the site. While I don't necessarily believe "build GraphQL support into Jekyll" makes sense, I do think Jekyll could provide more best-practices advice, tooling, and examples for how to build sites out via APIs. Personally, I actually prefer keeping data in Markdown and YML/JSON files in the repo as much as possible, so this isn't even a feature request I'm super excited about…more something I see Jekyll simply needing in order to remain relevant in the face of competing SSGs.

* **Easy Multilingual Setup**. Supporting content in multiple languages is unforutnately not a strength of Jekyll, but I've seen other SSGs (such as Next.js) also struggle with this, so it's not a straightforward problem to solve by any means. Nevertheless, I think it's something that Jekyll should try to improve on out-of-the-box as much as possible, or at least provide the necessary hooks under-the-hood so extra Jekyll plugins can solve the problem more easily.

Anyway, these are just a few of my thoughts, and I want to be clear I'm more than willing to help draft any necessary documentation or plans or whatever that could further these or other future goals the community comes up with. I might be able to work on some code PRs to submit down the line as well, since I've made it a personal goal in recent months to become more familiar with Jekyll's internals.

Comments? Ideas? Suggestions?

**Comment by DirtyF** [comment]
As you might have noticed, Jekyll development is mostly in a maintenance phase, so I don't see any of this happen anytime soon by magic, except if the Ruby community wants to help with plugins. Maybe you can? Or maybe you need to start thinking if another tooling does not fit your current requirements better?

**Comment by jaredcwhite** [comment]
Wow, OK. Well I would say that is a marketing problem not a coding problem. I've seen a lot of open source projects come and go and the ones that stick around have a core group of cheerleaders…updating docs, posting blog articles, regularly requesting community involvement and feedback, etc. All that to say, I think maybe a good place to start would be the Jekyll blog. Last post was from August last year! Is that something I could help draft up?

**Comment by ashmaroli** [comment]
@jaredcwhite Thank you for starting this discussion.

I don't think Jekyll will ever have an interface to bundle front-end packages (unless the community comes up with a really awesome solution). So, there won't be an officially sanctioned `jekyll-webpack` or similar, mainly because that is not what Jekyll is meant for.

Jekyll Core has just one goal: **Convert a bunch of text files into a website that is navigable when deployed by a web-server.**
Everything else is plugin-territory.

---

That said, Jekyll already has a rudimentary interface for *component* based development.
Place components in the `_includes` directory and *import* them into your layouts and documents via `{% include ... %}`.

---

Multilingual set up is something our users have asked for a long time.
We're not against the idea.., just don't see how it fits in the current Core architecture without significantly impacting performance.

**Comment by jaredcwhite** [comment]
I appreciate the response @ashmaroli. Again, I'm feeling like we're talking at cross purposes a bit here. You don't need to convince me I "can" do all these different things now with plugins and various tweaks. I'm already doing it. 😃

What I'm talking about is the perception Jekyll has within the outside world. The last blog post is over six months old. Nothing on the Jekyll homepage or intro Docs page highlights the major draws modern web devs are looking for (some of which I've presented). The list of Jekyll plugins/extensions includes links to abandonware that hasn't been touched in nearly a decade. And the answer to "is there a future roadmap?" seems to be, effectively, _nope_.

In short, it "looks" like Jekyll is just a tool for hosting simple blogs on GitHub and is slowly dying—both statements I believe are quite false. (i.e. there are plenty of advanced users of Jekyll out there building large complicated websites and hosting them on fancy platforms like Netlify. I should know because I'm one of those users!)

So even if I completely agree with you that Jekyll Core should stay entirely focused on its historical sweet spot and leave everything else to third-party plugins, that doesn't change the fact that potential new users looking at Jekyll's website and then looking at Gatsby, Next, Eleventy, Hugo, etc. are going to see a stark contrast. Jekyll is old and busted, these other tools are the new hotness. Again: I don't believe that's actually the case! But it's all about perception.

Anyway, I've yammered on enough for now. Like I said, I'd be very willing and able to help freshen up the website and roadmap docs and anything else that might trigger some more community involvement, if that's something you'd be interested in.

**Comment by ashmaroli** [comment]
@jaredcwhite We're always open to things that would trigger more community involvement.
You're welcome to submit pull requests to that effect or if you're comfortable with Ruby, to implement above functionality or help resolve some of the issues reported in this repository.

Regarding the roadmap, to be honest, there isn't one per se. (Yeah, you said that already). If you go through our bug-tracker, you'll see that there are many of them tagged with `has-pull-request`. You can go through the `History.markdown` document to see what's been changed so far (since the last release), and so on..

The bigger picture is that Jekyll's hotness was tied to GitHub Pages. Since now a user can easily use GitHub Pages' competitors to deploy a Jekyll site with a gemset (or an entirely different stack) of the user's choice, GitHub Pages isn't too keen on investing resources on testing out new functionality that may have an unforeseen effect on their users that rely on established behavior.
Something along the lines of &mdash; *"If it ain't broke, don't fix it"*.

In short, while there won't be an official post until the next release, you're welcome to engage the community with teasers based on this repo's current state and submit pull requests to improve our website.

**Comment by jaredcwhite** [comment]
Thanks @ashmaroli and @DirtyF for your views. I have to be honest—at this point I'm going to have to look further afield for a technology stack I can bet the future of my business on. All the best to you and the Jekyll team.
